File: core/matchers/src/main/scala/au/com/dius/pact/matchers/QueryMatcher.scala
package au.com.dius.pact.matchers

import au.com.dius.pact.core.model.matchingrules.MatchingRules
import au.com.dius.pact.model.{QueryMismatch, QueryMismatchFactory}
import com.typesafe.scalalogging.StrictLogging

import scala.collection.JavaConverters._

object QueryMatcher extends StrictLogging {

  def compare(parameter: String, path: Seq[String], expected: String, actual: String, matchers: MatchingRules) = {
    if (Matchers.matcherDefined("query", Seq(parameter).asJava, matchers)) {
      logger.debug("compareQueryParameterValues: Matcher defined for query parameter " + parameter)
      Matchers.domatch[QueryMismatch](matchers, "query", Seq(parameter).asJava, expected, actual, QueryMismatchFactory).asScala.toList
    } else {
      logger.debug("compareQueryParameterValues: No matcher defined for query parameter " + parameter + ", using equality")
      if (expected == actual) {
        Seq[QueryMismatch]()
      } else {
        Seq(QueryMismatch(parameter, expected, actual, Some(s"Expected '$expected' but received '$actual' for query parameter '$parameter'"),
          parameter))
      }
    }
  }

  def compareQueryParameterValues(parameter: String, expected: List[String], actual: List[String],
                                  path: Seq[String], matchers: MatchingRules) = {
    var result = Seq[QueryMismatch]()
    for ((value, index) <- expected.view.zipWithIndex) {
      if (index < actual.size) {
        result = result ++: compare(parameter, path :+ index.toString, value, actual(index), matchers)
      } else if (!Matchers.matcherDefined("query", path.asJava, matchers)) {
        result = result :+ QueryMismatch(parameter, expected.toString(), actual.toString(),
          Some(s"Expected query parameter $parameter value $value but was missing"),
          path.mkString("."))
      }
    }
    result
  }

  def compareQuery(parameter: String, expected: List[String], actual: List[String], matchers: MatchingRules) = {
    var result = Seq[QueryMismatch]()
    val path = Seq(parameter)
    if (Matchers.matcherDefined("query", path.asJava, matchers)) {
      logger.debug("compareQuery: Matcher defined for query parameter " + parameter)
      result = Matchers.domatch[QueryMismatch](matchers, "query", path.asJava, expected, actual, QueryMismatchFactory).asScala.toList ++
        compareQueryParameterValues(parameter, expected, actual, path, matchers)
    } else {
      if (expected.isEmpty && actual.nonEmpty) {
        result = Seq(QueryMismatch(parameter, expected.toString(), actual.toString(), Some(s"Expected an empty parameter List for $parameter but received $actual"),
          path.mkString(".")))
      } else {
        if (expected.size != actual.size) {
          result = result :+ QueryMismatch(parameter, expected.toString(), actual.toString(),
            Some(s"Expected query parameter $parameter with ${expected.size} values but received ${actual.size} values"),
            path.mkString("."))
        }
        result = result ++ compareQueryParameterValues(parameter, expected, actual, path, matchers)
      }
    }
    result
  }
}

File: core/matchers/src/main/kotlin/au/com/dius/pact/core/matchers/QueryMatcher.kt
package au.com.dius.pact.core.matchers

import mu.KLogging
import org.atteo.evo.inflector.English

object QueryMatcher : KLogging() {

  private fun compare(
    parameter: String,
    expected: String,
    actual: String,
    context: MatchingContext
  ): List<QueryMismatch> {
    return if (context.matcherDefined(listOf(parameter))) {
      logger.debug { "compareQueryParameterValues: Matcher defined for query parameter '$parameter'" }
      Matchers.domatch(context, listOf(parameter), expected, actual, QueryMismatchFactory)
    } else {
      logger
        .debug { "compareQueryParameterValues: No matcher defined for query parameter '$parameter', using equality" }
      if (expected == actual) {
        emptyList()
      } else {
        listOf(QueryMismatch(parameter, expected, actual, "Expected '$expected' but received '$actual' " +
          "for query parameter '$parameter'", parameter))
      }
    }
  }

  private fun compareQueryParameterValues(
    parameter: String,
    expected: List<String>,
    actual: List<String>,
    path: List<String>,
    context: MatchingContext
  ): List<QueryMismatch> {
    return expected.mapIndexed { index, value -> index to value }
      .flatMap { (index, value) ->
        when {
          index < actual.size -> compare(parameter, value, actual[index], context)
          !context.matcherDefined(path) ->
            listOf(QueryMismatch(parameter, expected.toString(), actual.toString(),
              "Expected query parameter '$parameter' with value '$value' but was missing",
              path.joinToString(".")))
          else -> emptyList()
        }
      }
  }

  @JvmStatic
  fun compareQuery(
    parameter: String,
    expected: List<String>,
    actual: List<String>,
    context: MatchingContext
  ): List<QueryMismatch> {
    val path = listOf(parameter)
    return if (context.matcherDefined(path)) {
      logger.debug { "compareQuery: Matcher defined for query parameter '$parameter'" }
      Matchers.domatch(context, path, expected, actual, QueryMismatchFactory) +
        compareQueryParameterValues(parameter, expected, actual, path, context)
    } else {
      if (expected.isEmpty() && actual.isNotEmpty()) {
        listOf(QueryMismatch(parameter, expected.toString(), actual.toString(),
          "Expected an empty parameter List for '$parameter' but received $actual",
          path.joinToString(".")))
      } else {
        val mismatches = mutableListOf<QueryMismatch>()
        if (expected.size != actual.size) {
          mismatches.add(QueryMismatch(parameter, expected.toString(), actual.toString(),
            "Expected query parameter '$parameter' with ${expected.size} " +
              "${English.plural("value", expected.size)} but received ${actual.size} " +
              English.plural("value", actual.size),
            path.joinToString(".")))
        }
        mismatches + compareQueryParameterValues(parameter, expected, actual, path, context)
      }
    }
  }
}
